File: src/algorithms/dynamic-programming/min-cost-climbing-stairs-memoization/sources/min-cost-climbing-stairs-memoization.py
# Min Cost Climbing Stairs memoization — top-down recursion with cached subproblems

def min_cost_climbing_stairs_memoization(costs: list[int]) -> int:
    memo: dict[int, int] = {}  # @step:initialize

    def compute_memo(step: int) -> int:
        if step <= 1:  # @step:initialize
            return 0  # @step:initialize
        if step in memo:  # @step:read-cache
            return memo[step]  # @step:read-cache
        # Recursively compute the minimum cost from each of the two preceding steps, cache to avoid recomputation
        # @step:push-call
        cost_from_one = compute_memo(step - 1) + costs[step - 1]  # @step:compute-cell
        cost_from_two = compute_memo(step - 2) + costs[step - 2]  # @step:compute-cell
        result = min(cost_from_one, cost_from_two)  # @step:compute-cell
        memo[step] = result  # @step:compute-cell
        # @step:pop-call
        return result  # @step:complete

    return compute_memo(len(costs))

File: src/algorithms/dynamic-programming/1d-linear/min-cost-climbing-stairs-memoization/sources/MinCostClimbingStairsMemoization.cpp
// Min Cost Climbing Stairs memoization — top-down recursion with cached subproblems

#include <iostream>
#include <vector>
#include <unordered_map>
#include <algorithm>

int computeMemo(const std::vector<int>& costs, int step, std::unordered_map<int, int>& memo) {
    if (step <= 1) return 0; // @step:initialize
    auto it = memo.find(step);
    if (it != memo.end()) return it->second; // @step:read-cache
    // Recursively compute the minimum cost from each of the two preceding steps, cache to avoid recomputation
    // @step:push-call
    int costFromOne = computeMemo(costs, step - 1, memo) + (step - 1 < (int)costs.size() ? costs[step - 1] : 0); // @step:compute-cell
    int costFromTwo = computeMemo(costs, step - 2, memo) + (step - 2 < (int)costs.size() ? costs[step - 2] : 0); // @step:compute-cell
    int result = std::min(costFromOne, costFromTwo); // @step:compute-cell
    memo[step] = result; // @step:compute-cell
    // @step:pop-call
    return result; // @step:complete
}

int minCostClimbingStairsMemoization(const std::vector<int>& costs) {
    // @step:initialize
    std::unordered_map<int, int> memo; // @step:initialize
    return computeMemo(costs, costs.size(), memo);
}

#ifndef TESTING
int main() {
    std::vector<int> costs = {10, 15, 20};
    int result = minCostClimbingStairsMemoization(costs);
    std::cout << "Min cost to climb: " << result << std::endl;
    return 0;
}
#endif
